src/services/impl/FileLogger.hh

```diff
@@ -0,0 +1,86 @@
+#ifndef RA_SERVICES_FILELOGGER_HH
+#define RA_SERVICES_FILELOGGER_HH
+#pragma once
+
+#include "services\IClock.hh"
+#include "services\IFileSystem.hh"
+#include "services\ILogger.hh"
+#include "services\ServiceLocator.hh"
+#include "services\TextWriter.hh"
+#include "services\impl\FileTextWriter.hh"
+
+#include <time.h>
+#include <mutex>
+
+namespace ra {
+namespace services {
+namespace impl {
+
+class FileLogger : public ra::services::ILogger
+{
+public:
+    explicit FileLogger(const ra::services::IFileSystem& pFileSystem) noexcept
+    {
+        m_pWriter = pFileSystem.AppendTextFile(pFileSystem.BaseDirectory() + L"RACache\\RALog.txt");
+        if (m_pWriter != nullptr)
+            m_pWriter->Write("\r\n");
+    }
+
+    bool IsEnabled([[maybe_unused]] LogLevel level) const override
+    {
+        return true;
+    }
+
+    void LogMessage(LogLevel level, const std::string& sMessage) const override
+    {
+        if (m_pWriter == nullptr)
+            return;
+
+        // write a timestamp
+        auto tNow = ServiceLocator::Get<IClock>().Now();
+        auto tMilliseconds = static_cast<unsigned int>(std::chrono::time_point_cast<std::chrono::milliseconds>(tNow).time_since_epoch().count() % 1000);
+        auto tTime = std::chrono::system_clock::to_time_t(tNow);
+
+        std::tm tTimeStruct;
+        localtime_s(&tTimeStruct, &tTime);
+
+        char sBuffer[16];
+        strftime(sBuffer, sizeof(sBuffer), "%H%M%S", &tTimeStruct);
+        sprintf_s(&sBuffer[6], sizeof(sBuffer) - 6, ".%03u|", tMilliseconds);
+
+        {
+            std::lock_guard<std::mutex> oLock(m_oMutex);
```

Comment:
I strongly suggest using `std::scoped_lock` instead. `std::lock_guard` should only be used for backwards-compatibility (which locks haven't been used before in this project). `std::scoped_lock `doesn't require specifying template arguments at all since they are deduced via perferct forwarding. It also does a better job at preventing deadlocks as it can lock more than one mutex at once.

How this one would look...
```
...
{
  std::scoped_lock oLock{m_oMutex};
...
```

Reply:
It is.